aptos-move/framework/src/natives/string_utils.rs

```diff
@@ -0,0 +1,463 @@
+
+struct FormatContext<'a, 'b, 'c, 'd> {
+    context: &'d mut SafeNativeContext<'a, 'b, 'c>,
+    base_gas: InternalGas,
+    max_depth: usize,
+    max_len: usize,
+    type_tag: bool,
+    canonicalize: bool,
+    single_line: bool,
+    include_int_type: bool,
+}
+
+/// Converts a `MoveValue::Vector` of `u8`'s to a `String` by wrapping it in double quotes and
+/// escaping double quotes and backslashes.
+///
+/// Examples:
+///  - 'Hello' returns "Hello"
+///  - '"Hello?" What are you saying?' returns "\"Hello?\" What are you saying?"
+///  - '\ and " are escaped' returns "\\ and \" are escaped"
+fn bytes_as_escaped_string(buf: &str) -> String {
+    let str = String::from(buf);
+
+    // We need to escape displayed double quotes " as \" and, as a result, also escape
+    // displayed \ as \\.
+    str.replace('\\', "\\\\").replace('"', "\\\"")
+}
+
+fn print_space_or_newline(newline: bool, out: &mut String, depth: usize) {
+    if newline {
+        out.push('\n');
+        for _ in 0..depth {
+            // add 2 spaces
+            write!(out, "  ").unwrap();
+        }
+    } else {
+        out.push(' ');
+    }
+}
+
+fn primitive_type(ty: &MoveTypeLayout) -> bool {
+    !matches!(ty, MoveTypeLayout::Vector(_) | MoveTypeLayout::Struct(_))
+}
+
+trait MoveLayout {
+    fn write_name(&self, out: &mut String);
+    fn get_layout(&self) -> &MoveTypeLayout;
+}
+
+impl MoveLayout for MoveFieldLayout {
+    fn write_name(&self, out: &mut String) {
+        write!(out, "{}: ", self.name).unwrap();
+    }
+
+    fn get_layout(&self) -> &MoveTypeLayout {
+        &self.layout
+    }
+}
+
+impl MoveLayout for MoveTypeLayout {
+    fn write_name(&self, _out: &mut String) {}
+
+    fn get_layout(&self) -> &MoveTypeLayout {
+        self
+    }
+}
+
+fn format_vector<'a>(
+    context: &mut FormatContext,
+    fields: impl Iterator<Item = &'a (impl MoveLayout + 'a)>,
+    values: Vec<Value>,
+    depth: usize,
+    newline: bool,
+    out: &mut String,
+) -> SafeNativeResult<()> {
+    if values.is_empty() {
+        return Ok(());
+    }
+    if depth >= context.max_depth {
+        write!(out, " .. ").unwrap();
+        return Ok(());
+    }
+    print_space_or_newline(newline, out, depth + 1);
+    for (i, (ty, val)) in fields.zip(values.into_iter()).enumerate() {
+        if i > 0 {
+            out.push(',');
+            print_space_or_newline(newline, out, depth + 1);
+        }
+        if i >= context.max_len {
+            write!(out, "..").unwrap();
+            break;
+        }
+        ty.write_name(out);
+        native_format_impl(context, ty.get_layout(), val, depth + 1, out)?;
+    }
+    print_space_or_newline(newline, out, depth);
+    Ok(())
+}
+
+fn native_format_impl(
+    context: &mut FormatContext,
+    layout: &MoveTypeLayout,
+    val: Value,
+    depth: usize,
+    out: &mut String,
+) -> SafeNativeResult<()> {
+    context.context.charge(context.base_gas)?;
```

Comment:
Do we need to charge gas based on the length of the output string? One potential problem I can think of here is I can parse a value with a very high number of nestings. This can lead to this function doing a lot more work than expected

Reply:
Well it's charging gas on all function entries, so nesting and repeated elements start charging gas at each entry of this function, so works automatically

Reply:
True we'll be charging base for each level of nesting. Is this sufficient to avoid execution time attacks though?

Reply:
Except for the special case of printing utf8 string all code path in this function are constant time so we only need to charge a constant amount each time